# End of Season / Phase 1 - Performance Evaluation

In [1]:
from utilities.results import Results
import pandas as pd
# used to show all rows in dfs
pd.set_option('display.max_rows', 500)

Phase 1 testing was the first real world testing of APPLE. Phase 1 commenced when the Premier League (PL) restarted on 17th June 2020 and finished 26th July 2020 when the 19/20 (PL)  season ended. 

Phase 1 had two primary objectives:
- asses APPLE’s real world predictive ability 
- build a workflow around the core APPLE models to automate as much as possible 

### Results table from Phase 1 testing

In [2]:
Results().running()

Fixture        Date APPLE Prediction DD Prediction  \
0   Aston Villa v Sheffield United  17/06/2020                A             A   
1               Man City v Arsenal  17/06/2020                H             H   
2            Norwich v Southampton  19/06/2020                A             H   
3           Tottenham v Man United  19/06/2020                H             A   
4              Watford v Leicester  20/06/2020                A             A   
5               Brighton v Arsenal  20/06.2020                D             D   
6                West Ham v Wolves  20/06.2020                A             A   
7     Bournemouth v Crystal Palace  20/06.2020                D             H   
8     Newcastle v Sheffield United  21/06/2020                A             D   
9            Aston Villa v Chelsea  21/06/2020                A             A   
10             Everton v Liverpool  21/06/2020                A             A   
11              Man City v Burnley  22/06/2020                H             H   
12            Leicester v Brighton  23/06/2020                H             H   
13            Tottneham v West Ham  23/06/2020                H             D   
14      Man Utd v Sheffield United  24/05/2020                D             H   
15         Newcastle v Aston Villa  24/05/2020                H             H   
16               Norwich v Everton  24/05/2020                A             A   
17            Wolves v Bournemouth  24/05/2020                H             H   
18      Liverpool v Crystal Palace  24/05/2020                H             H   
19               Burnley v Watford  25/06/2020                H             D   
20           Southampton v Arsenal  25/06/2020                A             D   
21              Chelsea v Man City  25/06/2020                A             D   
22            Aston Villa v Wolves  27/06/2020                A             A   
23           Watford v Southampton  28/06/2020                H             D   
24        Crystal Palace v Burnley  29/06/2020                D             H   
25              Brighton v Man Utd  30/06/2020                A             D   
26              Arsenal v Norwhich  2020-01-07                H             H   
27          Bournmouth v Newcastle  2020-01-07                H             A   
28             Everton v Leicester  2020-01-07                H             D   
29              West Ham v Chelsea  2020-01-07                A             A   
30    Sheffield United v Tottenham  2020-02-07                A             D   
31            Man City v Liverpool  2020-02-07                A             A   
32              Norwich v Brighton  2020-04-07                H             D   
33      Leicester v Crystal Palace  2020-04-07                A             H   
34           Man Utd v Bournemouth  2020-04-07                H             H   
35                Wolves v Arsenal  2020-04-07                D             H   
36               Chelsea v Watford  2020-04-07                H             H   
37      Burnley v Sheffield United  2020-05-07                A             D   
38            Newcastle v West Ham  2020-05-07                H             H   
39         Liverpool v Aston Villa  2020-05-07                H             H   
40          Southampton v Man City  2020-05-07                A             A   
41             Tottenham v Everton  2020-06-07                H             D   
42        Crystal Palace v Chelsea  2020-07-07                A             D   
43              Watford v Norwhich  2020-07-07                H             H   
44             Arsenal v Leicester  2020-07-07                H             H   
45            Man City v Newcastle  2020-08-07                H             H   
46       Sheffield United v Wolves  2020-08-07                A             D   
47              West Ham v Burnley  2020-08-07                H             D   
48            Brighton v Liverpool  2020-08-07            

Total Correct Predictions Accuracy of Total Predictions (%)
0     JR                        45                           54.2169
1  APPLE                        36                           43.3735
2     MW                        36                           43.3735
3     DD                        34                           40.9639

note: running results log needs some cleaning, DQ issues 

### APPLE Performance

- over 82 games, APPLE had a prediction accuracy of 43.3735%
- the best human predictor had a prediction accuracy of 54.2169

Consulting the log of saved model performance:

In [16]:
model_log = pd.read_csv("~/PycharmProjects/APPLE/saved_models/model_log.csv")
model_log = model_log.sort_values(by = ["Test Acc"], ascending = False).reset_index(drop = True)
display(model_log)

Model type      Date                              Model ID  Test Loss  \
0           2  20200717  f455690c-c85c-11ea-8321-acde48001122        NaN   
1           3  20200717  f655eb28-c85c-11ea-836e-acde48001122   0.932125   
2           3  20200725  584bd528-cea3-11ea-bd48-b29c4ab6f9cd   0.872550   
3           3  20200705  e19b479e-befe-11ea-ba1a-acde48001122   0.972137   
4           3  20200717  9c739ae2-c85b-11ea-a884-acde48001122   0.912533   
5           2  20200717  ccfc62d2-c85b-11ea-bc7f-acde48001122        NaN   
6           3  20200713  87412cdc-c531-11ea-a4f3-acde48001122   0.932609   
7           3  20200725  a5e84974-cea3-11ea-ac04-b29c4ab6f9cd   0.936698   

   Test Acc  
0  0.577465  
1  0.577465  
2  0.577465  
3  0.563380  
4  0.549296  
5  0.549296  
6  0.549296  
7  0.549296

The best performing model is a three way tie between ***f455690c-c85c-11ea-8321-acde48001122***, ***f655eb28-c85c-11ea-836e-acde48001122*** and ***584bd528-cea3-11ea-bd48-b29c4ab6f9cd*** with prediction accuracy of 57.7465%

This is considerably above the actual perfornace of 43.3735% and would have beaten the best human predictor at 54.2169%

There are two factors to consider at this point:
- APPLE was used from the restart of the Premier League (PL) after it had been suspended as a results of the COVID-19 pandemic. The games from the restart have been played have at an increased frequency and without crowds. It is conceivable that this has had a material, tangible impact om restuls. The models, initally, were trained on data from games played under "normal" circumstances
- Initally models were trained on data from before the suspension of the PL, part way through the restarted season backtesting with mined data was implemented and used to update test acc in the model log above. Thus a vareity of models were used throughout the season.

For the first bullet point it is difficult (if not impossible) to quantify impact but for the second bullet point we could backtest the top performing models to see if they had been used they would have beaten the top performing human predictor

There is however, an important point to raise here: the best perfomring models are those with the most data. The date column in the table above is when these models were trained and saved. Over the course of phase 1, 3 models were built each week. Each model was then backtested on mined data and it's performance saved. The saved models were then deleted if they were underprofming on backtesting data. This means that the 7 retained models are from July. They perform well on backtesting becuase they were trained on the same data. This is a fallacy in my apporach and will be removed from V2 APPLE. 

## Next Steps

Given the lessons learned from Phase 1 the next steps are:
- refactor the entire project to improve code and structure of entire project
- deep dive on core models used for prediction, asses whether there are better features to be used, this includes looking for new data sources 
